### PR TITLE
Add WebAssembly compilation

### PR DESCRIPTION
**Pull request by vinciusb**
Augment the build system allowing it to compile cvc5 to WebAssembly. This built-in feature uses [emscripten](https://emscripten.org/) (as an external dependency) which offers 3 possible compilation extensions:
- WASM: only generates the cvc5.wasm.
- JS: generate cvc5.wasm and a glue code (.js).
- HTML: generate cvc5.wasm, a glue code (.js) and a web page (.html).

> Signed-off-by: Vinícius Braga Freire [vinicius.braga.freire@gmail.com](mailto:vinicius.braga.freire@gmail.com)

### CODE REVIEW

File: cmake/FindGMP.cmake [hunk 1]
```diff
@@ -113,11 +113,12 @@ if(NOT GMP_FOUND_SYSTEM)
     URL_HASH SHA1=2dcf34d4a432dbe6cce1475a835d20fe44f75822
     CONFIGURE_COMMAND
       ${CONFIGURE_ENV}
-        <SOURCE_DIR>/configure
+          ${CONFIGURE_CMD_WRAPPER} <SOURCE_DIR>/configure
           ${LINK_OPTS}
           --prefix=<INSTALL_DIR>
           --with-pic
           --enable-cxx
+
```
Comment (vinciusb) [inline cmake/FindGMP.cmake:121]:
```suggestion
```

File: cmake/FindPoly.cmake [hunk 2]
```diff
@@ -143,6 +143,17 @@ if(NOT Poly_FOUND_SYSTEM)
...
   endif()
 
+  # If it isn't a wasm compilation
+  if(WASM STREQUAL "OFF")
+    set(POLY_PATCH_CMD_WASM "")
+  # Otherwise
+  else()
+    set(POLY_PATCH_CMD_WASM ${POLY_PATCH_CMD_WASM}
+      COMMAND
+        bash ${CMAKE_SOURCE_DIR}/cmake/deps-utils/Poly-wasm-patch.sh ${DEPS_BASE}
```
Comment (4tXJ7f) [inline cmake/FindPoly.cmake:153]: A patch file would be better than a script in this case (see [here](https://github.com/cvc5/cvc5/blob/d01e59c13b7337e68806be72145b3e45e4cad89c/cmake/FindSymFPU.cmake#L42-L43) for an example).

File: cmake/FindPoly.cmake [hunk 3]
```diff
@@ -143,6 +143,17 @@ if(NOT Poly_FOUND_SYSTEM)
       "${DEPS_BASE}/lib/libpicpolyxx${CMAKE_STATIC_LIBRARY_SUFFIX}")
   endif()
 
+  # If it isn't a wasm compilation
+  if(WASM STREQUAL "OFF")
+    set(POLY_PATCH_CMD_WASM "")
```
Comment (4tXJ7f) [inline cmake/FindPoly.cmake:148]: I think that it would be cleaner/easier to just append to `POLY_PATCH_CMD` as needed instead of introducing `POLY_PATCH_CMD_WASM`.

File: cmake/FindGMP.cmake (hunk 1 above)
Comment (4tXJ7f) [reply cmake/FindGMP.cmake:121]: Apply? :)

File: configure.sh [hunk 4]
```diff
@@ -134,6 +138,9 @@ ipo=default
 
 glpk_dir=default
 
+wasm=OFF
```
Comment (4tXJ7f) [inline configure.sh:141]:
```suggestion
wasm=default
```

To be consistent with the other options. Also, below, we are checking `"$wasm" != default`, which can only be false if we set `wasm` to `default` here.

File: src/main/CMakeLists.txt [hunk 5]
```diff
@@ -58,11 +58,25 @@ endif()
...
+string(REGEX REPLACE "-s +" "-s" WASM_FLAGS "${WASM_FLAGS}")
+
 add_executable(cvc5-bin driver_unified.cpp main.cpp $<TARGET_OBJECTS:main>)
 target_compile_definitions(cvc5-bin PRIVATE -D__BUILDING_CVC5DRIVER -Dcvc5_obj_EXPORTS)
+target_link_options(cvc5-bin PRIVATE ${WASM_FLAGS})
 set_target_properties(cvc5-bin
   PROPERTIES
-    OUTPUT_NAME cvc5
+    OUTPUT_NAME cvc5 
+    SUFFIX "${BINARY_EXTENSION}"
```
Comment (4tXJ7f) [inline src/main/CMakeLists.txt:79]: That seems wrong when we compile for Windows and need the `.exe` extension. Instead, we should set `CMAKE_EXECUTABLE_SUFFIX` in the root [CMakeLists.txt file](https://github.com/cvc5/cvc5/blob/6bd7888f807584f02c52821d4c5d69c383dda22e/CMakeLists.txt) (so that all executable files have the correct extension).

File: src/main/CMakeLists.txt [hunk 6]
```diff
@@ -58,11 +58,25 @@ endif()
...
+if(WASM STREQUAL "WASM")
+  set(BINARY_EXTENSION ".wasm")
+elseif(WASM STREQUAL "JS")
+  set(BINARY_EXTENSION ".js")
+elseif(WASM STREQUAL "HTML")
+  set(BINARY_EXTENSION ".html")
+else()
+  set(BINARY_EXTENSION "")
+endif()
+string(REGEX REPLACE "-s +" "-s" WASM_FLAGS "${WASM_FLAGS}")
```
Comment (4tXJ7f) [inline src/main/CMakeLists.txt:71]: Why is this necessary?

File: cmake/Helpers.cmake [hunk 7]
```diff
@@ -146,6 +146,16 @@ macro(cvc5_option var description)
...
 endmacro()
 
+# Wasm options are four-valued to detect if the WebAssembly compilation will
+# happen and if so, find out the extension used in this process.
+# The options are: OFF (not a WebAssembly compilation), WASM (.wasm), JS (.js) 
+# or HTML (.html).
+macro(wasm_option var description)
+  set(${var} IGNORE CACHE STRING "${description}")
+  # Provide drop down menu options in cmake-gui
+  set_property(CACHE ${var} PROPERTY STRINGS OFF WASM JS HTML)
```
Comment (4tXJ7f) [inline cmake/Helpers.cmake:156]: I wouldn't make this a separate macro, because it only applies to one specific option. Do we do that for any other option?

File: CMakeLists.txt [hunk 8]
```diff
@@ -128,6 +128,28 @@ option(BUILD_DOCS "Build Api documentation")
 cvc5_option(STATIC_BINARY "Link against static system libraries \
 (enabled by default for static builds)")
 
+# Custom extension for WebAssembly compilation
```
Comment (4tXJ7f) [inline CMakeLists.txt:131]:
```suggestion
# Custom flags for WebAssembly compilation
```

File: CMakeLists.txt [hunk 9]
```diff
@@ -128,6 +128,28 @@ option(BUILD_DOCS "Build Api documentation")
...
+set(WASM_FLAGS "" CACHE STRING "Link flags for the WebAssembly binary generation")
+wasm_option(WASM "Use a specific extension for WebAssembly compilation")
+
+# Make sure no other avaible value will be assigned to this variable
+if(NOT(WASM STREQUAL "WASM" OR WASM STREQUAL "JS" OR WASM STREQUAL "HTML"))
+  set(WASM "OFF")
```
Comment (4tXJ7f) [inline CMakeLists.txt:137]: Instead of silently changing to `OFF`, we should just throw an error. Makes it easier to catch potential typos.

File: CMakeLists.txt [hunk 10]
```diff
@@ -128,6 +128,28 @@ option(BUILD_DOCS "Build Api documentation")
...
+# If it's a wasm compilation, but isn't using static libraries
+elseif(BUILD_SHARED_LIBS OR NOT(STATIC_BINARY))
+  message(FATAL_ERROR "Use --static and --static-binary flags for WebAssembly compilation. GMP will not compile with shared libraries.")
+endif()
+
+# Not WebAssembly compilation
+if(WASM STREQUAL "OFF")
+  set(WASM_FLAGS "")
+  set(CONFIGURE_CMD_WRAPPER "")
+# Then the WebAssembly compilation will happen
```
Comment (4tXJ7f) [inline CMakeLists.txt:147]:
```suggestion
```

File: CMakeLists.txt [hunk 11]
```diff
@@ -128,6 +128,28 @@ option(BUILD_DOCS "Build Api documentation")
...
+
+# Make sure no other avaible value will be assigned to this variable
+if(NOT(WASM STREQUAL "WASM" OR WASM STREQUAL "JS" OR WASM STREQUAL "HTML"))
+  set(WASM "OFF")
+# If it's a wasm compilation, but isn't using static libraries
+elseif(BUILD_SHARED_LIBS OR NOT(STATIC_BINARY))
+  message(FATAL_ERROR "Use --static and --static-binary flags for WebAssembly compilation. GMP will not compile with shared libraries.")
+endif()
+
+# Not WebAssembly compilation
```
Comment (4tXJ7f) [inline CMakeLists.txt:143]: The if statement is self-explanatory.

```suggestion
```

File: INSTALL.rst [hunk 12]
```diff
@@ -54,6 +54,37 @@ The built binary ``cvc5.exe`` is located in ``<build_dir>/bin`` and the cvc5
 library can be found in ``<build_dir>/lib``.
 
 
+WebAssembly Compilation
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+Compiling cvc5 to WebAssembly needs the emscripten SDK (version 3.1.18). Setting
```
Comment (4tXJ7f) [inline INSTALL.rst:60]:
```suggestion
Compiling cvc5 to WebAssembly needs the Emscripten SDK (version 3.1.18). Setting
```

Let's say `version 3.1.18 or later`.

File: INSTALL.rst [hunk 13]
```diff
@@ -54,6 +54,37 @@ The built binary ``cvc5.exe`` is located in ``<build_dir>/bin`` and the cvc5
...
+
+.. code:: bash
+
+  git clone https://github.com/emscripten-core/emsdk.git
+  cd emsdk
+  ./emsdk install <version>   # <version> = '3.1.18' is preferable, but 
+                              # <version> = 'latest' has high chance of working
+  ./emsdk activate <version>
+  source ./emsdk_env.sh   # Activate PATH and other environment variables in the
+                          # current terminal. Whenever emscripten is going to be
```
Comment (4tXJ7f) [inline INSTALL.rst:71]:
```suggestion
                          # current terminal. Whenever Emscripten is going to be
```

File: INSTALL.rst [hunk 14]
```diff
@@ -54,6 +54,37 @@ The built binary ``cvc5.exe`` is located in ``<build_dir>/bin`` and the cvc5
 library can be found in ``<build_dir>/lib``.
 
 
+WebAssembly Compilation
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+Compiling cvc5 to WebAssembly needs the emscripten SDK (version 3.1.18). Setting
+ up EMSDK can be done as follows:
```
Comment (4tXJ7f) [inline INSTALL.rst:61]:
```suggestion
 up emsdk can be done as follows:
```

File: INSTALL.rst [hunk 15]
```diff
@@ -54,6 +54,37 @@ The built binary ``cvc5.exe`` is located in ``<build_dir>/bin`` and the cvc5
...
+                              # <version> = 'latest' has high chance of working
+  ./emsdk activate <version>
+  source ./emsdk_env.sh   # Activate PATH and other environment variables in the
+                          # current terminal. Whenever emscripten is going to be
+                          # used this command needs to be called before because 
+                          # emsdk doesn't insert the binaries paths directly in 
+                          # the system PATH variable.
+
+Look at the `emscripten dependencies list <https://emscripten.org/docs/getting_started/downloads.html#platform-specific-notes>`_
+. The user is expected to install it beforehand with his own system package manager.
```
Comment (4tXJ7f) [inline INSTALL.rst:77]:
```suggestion
Refer to the `emscripten dependencies list <https://emscripten.org/docs/getting_started/downloads.html#platform-specific-notes>`_ to ensure that all required dependencies are installed on the system.
```

**Review by 4tXJ7f — CHANGES_REQUESTED** [review]
It is exciting to have that on `main` soon! Just some minor requests.

File: cmake/FindPoly.cmake (hunk 2 above)
Comment (vinciusb) [reply cmake/FindPoly.cmake:153]: More than that, I will remove the need for any bash/patch file. I will insert an emscripten compilation flag that avoids the problem that this patch file is supposed to resolve. This approach is better because it makes sure we will have no problem with the poly lib in the future if they start using the variable that this patch comments.

File: configure.sh [hunk 16]
```diff
@@ -69,6 +69,10 @@ Optional Path to Optional Packages:
 CMake Options (Advanced)
   -DVAR=VALUE              manually add CMake options
 
+Wasm Options
+  --wasm=VALUE             set compilation extension for WebAssembly <OFF, WASM, JS or HTML>
+  --wasm-flags=STR         emscripten flags used in the WebAssembly binary compilation
```
Comment (vinciusb) [inline configure.sh:74]:
```suggestion
  --wasm-flags='STR'       emscripten flags used in the WebAssembly binary compilation
```

File: cmake/Helpers.cmake (hunk 7 above)
Comment (vinciusb) [reply cmake/Helpers.cmake:156]: It doesn't happen in any other place. I will fix it.

File: src/main/CMakeLists.txt (hunk 6 above)
Comment (vinciusb) [reply src/main/CMakeLists.txt:71]: It isn't. I was using it to standardize the flags being passed to the link process. I will remove it, didn't realize that was useless.

File: cmake/FindPoly.cmake [hunk 17]
```diff
@@ -143,6 +143,14 @@ if(NOT Poly_FOUND_SYSTEM)
...
   endif()
 
+  if(NOT(WASM STREQUAL "OFF"))
+    set(POLY_PATCH_CMD ${POLY_PATCH_CMD}
+      COMMAND
+        patch -i ${CMAKE_SOURCE_DIR}/cmake/deps-utils/Poly-wasm.patch 
+                 ${DEPS_BASE}/src/Poly-EP/src/upolynomial/factorization.c
+    )
+  endif()
+
```
Comment (vinciusb) [inline cmake/FindPoly.cmake:153]: Try to substitute this patch by inserting a global flag that avoids the abortion of the compilation process when detecting `declared but unused variables`.

File: INSTALL.rst [hunk 18]
```diff
@@ -54,6 +54,37 @@ The built binary ``cvc5.exe`` is located in ``<build_dir>/bin`` and the cvc5
...
+                          # the system PATH variable.
+
+Refer to the `emscripten dependencies list <https://emscripten.org/docs/getting_started/downloads.html#platform-specific-notes>`_ 
+to ensure that all required dependencies are installed on the system.
+
+Then, in the cvc5:
+
+.. code:: bash
+
+  ./configure.sh --static --static-binary --auto-download --wasm=<value> <configure options...>
```
Comment (vinciusb) [inline INSTALL.rst:83]:
```suggestion
  ./configure.sh --static --static-binary --auto-download --wasm=<value> --wasm-flags='<value>' <configure options...>
```

File: INSTALL.rst [hunk 19]
```diff
@@ -54,6 +54,37 @@ The built binary ``cvc5.exe`` is located in ``<build_dir>/bin`` and the cvc5
...
+                          # the system PATH variable.
+
+Refer to the `emscripten dependencies list <https://emscripten.org/docs/getting_started/downloads.html#platform-specific-notes>`_ 
+to ensure that all required dependencies are installed on the system.
+
+Then, in the cvc5:
+
+.. code:: bash
+
+  ./configure.sh --static --static-binary --auto-download --wasm=<value> --wasm-flags='<value>' <configure options...>
```
Comment (vinciusb) [inline INSTALL.rst:83]:
```suggestion
  ./configure.sh --static --static-binary --auto-download --wasm=<value> --wasm-flags='<emscripten flags>' <configure options...>
```

File: INSTALL.rst [hunk 20]
```diff
@@ -54,6 +54,56 @@ The built binary ``cvc5.exe`` is located in ``<build_dir>/bin`` and the cvc5
...
+  source ./emsdk_env.sh   # Activate PATH and other environment variables in the
+                          # current terminal. Whenever Emscripten is going to be
+                          # used this command needs to be called before because 
+                          # emsdk doesn't insert the binaries paths directly in 
+                          # the system PATH variable.
+
+Refer to the `emscripten dependencies list <https://emscripten.org/docs/getting_started/downloads.html#platform-specific-notes>`_ 
+to ensure that all required dependencies are installed on the system.
+
+Then, in the cvc5:
```
Comment (4tXJ7f) [inline INSTALL.rst:79]:
```suggestion
Then, in the cvc5 directory:
```

File: INSTALL.rst [hunk 21]
```diff
@@ -54,6 +54,56 @@ The built binary ``cvc5.exe`` is located in ``<build_dir>/bin`` and the cvc5
...
+(not only the wasm, but the .js glue code for web integration) and ``HTML`` (both
+the last two files and also an .html file which supports the run of the glue
+code).
+
+``--wasm-flags`` take a string wrapped by a single quote containing the
+`emscripten flags <https://github.com/emscripten-core/emscripten/blob/main/src/settings.js>`_,
+which modifies how the wasm and glue code are built and how they behave. An ``-s``
+should precede each flag.
+
+For an example if one wants to have an modularized glue code, you can try:
```
Comment (4tXJ7f) [inline INSTALL.rst:98]:
```suggestion
For example, to generate modularized glue code, use:
```

File: cmake/FindPoly.cmake [hunk 22]
```diff
@@ -143,6 +143,14 @@ if(NOT Poly_FOUND_SYSTEM)
       "${DEPS_BASE}/lib/libpicpolyxx${CMAKE_STATIC_LIBRARY_SUFFIX}")
   endif()
 
+  set(POLY_CC_FLAGS )
+  if(NOT(WASM STREQUAL "OFF"))
+    set(POLY_CC_FLAGS 
+          -DCMAKE_CXX_FLAGS=-Wno-error=unused-but-set-variable
+          -DCMAKE_C_FLAGS=-Wno-error=unused-but-set-variable
+        )
```
Comment (4tXJ7f) [inline cmake/FindPoly.cmake:151]: Can you add a comment that describes why this warning suppression is required (with maybe a link to the affected code)? This will make it easier to remove this special case in the future when it is not needed anymore.

File: configure.sh [hunk 23]
```diff
@@ -69,6 +69,10 @@ Optional Path to Optional Packages:
 CMake Options (Advanced)
   -DVAR=VALUE              manually add CMake options
 
+Wasm Options
+  --wasm=VALUE             set compilation extension for WebAssembly <OFF, WASM, JS or HTML>
```
Comment (4tXJ7f) [inline configure.sh:73]:
```suggestion
  --wasm=VALUE             set compilation extension for WebAssembly <WASM, JS or HTML>
```

File: configure.sh [hunk 24]
```diff
@@ -69,6 +69,10 @@ Optional Path to Optional Packages:
...
+  --wasm-flags='STR'       emscripten flags used in the WebAssembly binary compilation
```
Comment (4tXJ7f) [inline configure.sh:74]:
```suggestion
  --wasm-flags='STR'       Emscripten flags used in the WebAssembly binary compilation
```

File: src/main/CMakeLists.txt [hunk 25]
```diff
@@ -60,6 +60,7 @@ endif()
 
 add_executable(cvc5-bin driver_unified.cpp main.cpp $<TARGET_OBJECTS:main>)
 target_compile_definitions(cvc5-bin PRIVATE -D__BUILDING_CVC5DRIVER -Dcvc5_obj_EXPORTS)
+target_link_options(cvc5-bin PRIVATE ${WASM_FLAGS})
```
Comment (4tXJ7f) [inline src/main/CMakeLists.txt:63]: Instead of modifying the linker options for only this binary, let's do it globally for all binaries like we do for other linker options: https://github.com/cvc5/cvc5/blob/6791c74f34495d6bc3f2f097a281d82835bf9808/CMakeLists.txt#L216

**Review by 4tXJ7f — CHANGES_REQUESTED** [review]
Thank you for all the updates. I just have some minor requests/suggestions left. After that we can finally merge :)

File: CMakeLists.txt [hunk 26]
```diff
@@ -132,6 +132,28 @@ option(BUILD_DOCS "Build Api documentation")
...
 (enabled by default for static builds)")
 
+# Custom flags for WebAssembly compilation
+set(WASM_FLAGS "" CACHE STRING "Link flags for the WebAssembly binary generation")
+set(WASM "OFF" CACHE STRING "Use a specific extension for WebAssembly compilation")
+
+# Make sure no other avaible value will be assigned to this variable
+if(NOT(WASM STREQUAL "WASM" OR WASM STREQUAL "JS" OR WASM STREQUAL "HTML" OR WASM STREQUAL "OFF"))
+  message(FATAL_ERROR "${WASM} is not a valid wasm flag value.")
+endif()
```
Comment (mpreiner) [inline CMakeLists.txt:142]: You should be able to define the set of possible values as follows (untested):
```suggestion
set_property(CACHE WASM PROPERTY STRINGS WASM JS HTML OFF)
```

File: CMakeLists.txt [hunk 27]
```diff
@@ -132,6 +132,28 @@ option(BUILD_DOCS "Build Api documentation")
 cvc5_option(STATIC_BINARY "Link against static system libraries \
 (enabled by default for static builds)")
 
+# Custom flags for WebAssembly compilation
+set(WASM_FLAGS "" CACHE STRING "Link flags for the WebAssembly binary generation")
```
Comment (mpreiner) [inline CMakeLists.txt:136]: When do we need these flags and where are they set? Is this purely a user option?

Comment (vinciusb) [reply CMakeLists.txt:136]: We need these flags when compiling into Web Assembly and need to pass some Emscripten flags to the compilation link. It's a user option so it is set in the configure.sh. The user needs to pass it as a string wrapped by single quote just like that: --wasm-flags='-s INCOMING_MODULE_JS_API=arguments -s INVOKE_RUN=1 -s MODULARIZE -s EXIT_RUNTIME=0 -s ENVIRONMENT=web'.

These flags modify how the wasm, glue code and HTML file are generated and behave.

File: CMakeLists.txt (hunk 26 above)
Comment (vinciusb) [reply CMakeLists.txt:142]: I tested and set_property apparently doesn't execute tests to check if the variable is in the desired domain. One way to approach that way is something like that:

```
set_property(CACHE WASM PROPERTY STRINGS WASM JS HTML OFF)
get_property(WASM_STRINGS CACHE WASM PROPERTY STRINGS)

# Make sure no other avaible value will be assigned to this variable
if(NOT WASM IN_LIST WASM_STRINGS)
  message(FATAL_ERROR "${WASM} is not a valid wasm flag value.")
endif()
```

File: CMakeLists.txt [hunk 28]
```diff
@@ -132,6 +132,28 @@ option(BUILD_DOCS "Build Api documentation")
...
+set(WASM "OFF" CACHE STRING "Use a specific extension for WebAssembly compilation")
+
+# Make sure no other avaible value will be assigned to this variable
```
Comment (vinciusb) [inline CMakeLists.txt:139]:
```suggestion
# Make sure no other available value will be assigned to this variable
```

File: cmake/FindPoly.cmake [hunk 29]
```diff
@@ -143,6 +143,20 @@ if(NOT Poly_FOUND_SYSTEM)
       "${DEPS_BASE}/lib/libpicpolyxx${CMAKE_STATIC_LIBRARY_SUFFIX}")
   endif()
 
+  set(POLY_CC_FLAGS )
+  if(NOT(WASM STREQUAL "OFF"))
+    # The flag -Wall is set as default in cvc5, then these flags bellow make
+    # sure the wasm compilation doesn't abort when compiling LibPoly. At
+    # https://github.com/SRI-CSL/libpoly/blob/master/src/upolynomial/factorization.c#L1269-L1274
+    # LibPoly have a variable called enabled_count that is instantiated but its
```
Comment (4tXJ7f) [inline cmake/FindPoly.cmake:151]:
```suggestion
    # LibPoly has a variable called enabled_count that is instantiated but its
```

File: cmake/FindPoly.cmake [hunk 30]
```diff
@@ -143,6 +143,20 @@ if(NOT Poly_FOUND_SYSTEM)
       "${DEPS_BASE}/lib/libpicpolyxx${CMAKE_STATIC_LIBRARY_SUFFIX}")
   endif()
 
+  set(POLY_CC_FLAGS )
+  if(NOT(WASM STREQUAL "OFF"))
+    # The flag -Wall is set as default in cvc5, then these flags bellow make
+    # sure the wasm compilation doesn't abort when compiling LibPoly. At
+    # https://github.com/SRI-CSL/libpoly/blob/master/src/upolynomial/factorization.c#L1269-L1274
```
Comment (4tXJ7f) [inline cmake/FindPoly.cmake:150]:
```suggestion
    # https://github.com/SRI-CSL/libpoly/blob/d2cc42c492d05b90069e20ba813ee32fef086566/src/upolynomial/factorization.c#L1269-L1274
```

It is better to link to a specific commit in case that the code gets updated.

**Review by 4tXJ7f — APPROVED** [review]
LGTM, thank you for all the updates!
